drivers/lora/rylrxxx.c

```diff
@@ -0,0 +1,684 @@
+};
+
+struct rylr_recv_msg {
+	uint16_t addr;
+	uint8_t length;
+	uint8_t *data;
+	uint8_t rssi;
+	uint8_t snr;
+};
+
+static void on_ok(struct modem_chat *chat, char **argv, uint16_t argc, void *user_data)
+{
+	int err = 0;
+	struct rylr_data *driver_data = user_data;
+
+	driver_data->handler_error = err;
+}
+
+static void on_err(struct modem_chat *chat, char **argv, uint16_t argc, void *user_data)
+{
+	int radio_err = 0;
+	struct rylr_data *driver_data = user_data;
+
+	driver_data->handler_error = -EIO;
+
+	if (argc != 2) {
+		driver_data->handler_error = -EBADMSG;
+		LOG_ERR("malformed error message from radio");
+		return;
+	}
+
+	radio_err = atoi(argv[1]);
+	LOG_ERR("error from rylr: %d", radio_err);
+}
+
+static void on_rx(struct modem_chat *chat, char **argv, uint16_t argc, void *user_data)
+{
+	struct rylr_data *driver_data = user_data;
+	struct rylr_recv_msg msg;
+	int err = 0;
+
+	driver_data->handler_error = 0;
+
+	if (argc != 6) {
+		driver_data->handler_error = -EBADMSG;
+		return;
+	}
+
+	msg.addr = atoi(argv[1]);
+	msg.length = atoi(argv[2]);
+	msg.data = argv[3];
+	msg.rssi = atoi(argv[4]);
+	msg.snr = atoi(argv[5]);
+
+	if (RYLR_IS_RX_PENDING(driver_data->pending_async_flags)) {
+		driver_data->async_rx_cb(driver_data->dev, msg.data, msg.length, msg.rssi, msg.snr);
+	} else {
+		err = k_msgq_put(&driver_data->rx_msgq, &msg, K_NO_WAIT);
+		if (err != 0) {
+			LOG_ERR("error adding messgae to queue: %d", err);
+			driver_data->handler_error = err;
+		}
+	}
+}
+
+static void on_script_finished(struct modem_chat *chat, enum modem_chat_script_result result,
+			       void *user_data)
+{
+	struct rylr_data *driver_data = user_data;
+
+	if (RYLR_IS_TX_PENDING(driver_data->pending_async_flags)) {
+		RYLR_CLEAR_TX_PENDING(driver_data->pending_async_flags);
+		k_poll_signal_raise(driver_data->async_tx_signal, driver_data->handler_error);
+		k_sem_give(&driver_data->operation_sem);
+	}
+
+	k_sem_give(&driver_data->script_sem);
+}
+
+MODEM_CHAT_MATCH_DEFINE(ok_match, "+OK", "", on_ok);
+
+MODEM_CHAT_MATCHES_DEFINE(abort_matches, MODEM_CHAT_MATCH("+ERR=", "", on_err));
+
+MODEM_CHAT_MATCHES_DEFINE(unsol_matches, MODEM_CHAT_MATCH("+RCV=", ",", on_rx));
+
+MODEM_CHAT_SCRIPT_CMDS_DEFINE(ping_cmd, MODEM_CHAT_SCRIPT_CMD_RESP("AT\r\n", ok_match));
+
+MODEM_CHAT_SCRIPT_DEFINE(ping_script, ping_cmd, abort_matches, on_script_finished,
+			 CONFIG_RYLRXXX_RADIO_CMD_RESPONSE_TIMEOUT_MS);
+
+static void rylr_reset_dynamic_script(struct rylr_data *data)
+{
+	data->dynamic_chat.response_matches = &ok_match;
+	data->dynamic_chat.response_matches_size = 1;
+	data->dynamic_chat.timeout = 0;
+
+	data->dynamic_script.script_chats = &data->dynamic_chat;
+	data->dynamic_script.script_chats_size = 1;
+	data->dynamic_script.abort_matches = abort_matches;
+	data->dynamic_script.abort_matches_size = 1;
+	data->dynamic_script.callback = on_script_finished;
+	data->dynamic_script.timeout = CONFIG_RYLRXXX_RADIO_CMD_RESPONSE_TIMEOUT_MS;
```

Comment:
Not blocking or anything, just a suggestion
```suggestion
	data->dynamic_chat = (<cast to type of dynamic chat>) {
		.response_matches = &ok_match;
		.response_matches_size = 1;
		.script_chats = &data->dynamic_chat;
		.script_chats_size = 1;
		.abort_matches = abort_matches;
		.abort_matches_size = 1;
		.callback = on_script_finished;
		.timeout = CONFIG_RYLRXXX_RADIO_CMD_RESPONSE_TIMEOUT_MS;
};
```